[WIP] Build Envoy proxy with Istio filters and TPROXY support

rlenglet opened the pull request:
Build the Istio filters into the Envoy proxy: https://github.com/istio/proxy

Pull a quick-and-dirty implementation of the SOL_IP,IP_TRANSPARENT option setup on listen sockets:
https://github.com/rlenglet/envoy/tree/quick-and-dirty-transparent
This will be reverted when the proper, clean change is merged upstream into Envoy:
https://github.com/envoyproxy/envoy/pull/2719

Build Cilium-specific version of Istio's proxy Docker image.

Signed-off-by: Romain Lenglet <romain@covalent.io>

tgraf commented on envoy/BUILD:
```diff
@@ -27,8 +27,17 @@ envoy_cc_binary(
+
+        # Istio filters.
+        # Cf. https://github.com/istio/proxy/blob/master/src/envoy/BUILD#L23
+        "@istio_proxy//src/envoy/http/jwt_auth:http_filter_factory",
```
Does this mean we will always build all istio filters for every Cilium build? What if a user doesn't want to run the istio filters?

rlenglet replied:
Building is different from running. They won’t be running in Cilium’s node proxy because we won’t configure them. They are configured only in the Istio sidecar, by pilot-agent.

rlenglet replied:
And we’ll need those filters when we support Istio in a sidecar-less model.

tgraf reviewed (dismissed):
Why is the TPROXY replacement required in the first place? It looks like we do the exact same unconditional redirect.

rlenglet committed:
envoy: Build Istio Docker images

Build Istio filters into Envoy binary.
Split Envoy binary build targets for Cilium and Istio.
Only build the Istio filters into the binary used in Istio sidecar
proxy containers.
Define all PHONY Makefile targets.

Define Makefile targets to build Istio 0.7.0 proxy images:
docker-istio-proxy builds image cilium/istio_proxy:*
docker-istio-proxy-debug builds image cilium/istio_proxy_debug:*

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
envoy: Add README.Istio.md to specify Istio image build process

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
docs: Update Istio GSG to Istio 0.8.0 pre-release

Signed-off-by: Romain Lenglet <romain@covalent.io>
